internal/database/database.go

```
@@ -204,6 +217,85 @@ func (d *cosmosDBClient) GetLockClient() *LockClient {
 	return d.lockClient
 }
 
+func (d *cosmosDBClient) getBillingID(ctx context.Context, resourceID *azcorearm.ResourceID) (string, error) {
+	pk := NewPartitionKey(resourceID.SubscriptionID)
+
+	// Resource ID alone does not uniquely identify a billing document, but
+	// resource ID AND the absence of a deletion timestamp should be unique.
+	const query = "SELECT c.id FROM c WHERE STRINGEQUALS(c.resourceId, @resourceId, true) AND NOT IS_DEFINED(c.deletionTime)"
+	opt := azcosmos.QueryOptions{
+		PageSizeHint: 1,
+		QueryParameters: []azcosmos.QueryParameter{
+			{
+				Name:  "@resourceId",
+				Value: resourceID.String(),
+			},
+		},
+	}
+
+	queryPager := d.billing.NewQueryItemsPager(query, pk, &opt)
+
+	for queryPager.More() {
+		queryResponse, err := queryPager.NextPage(ctx)
+		if err != nil {
+			return "", fmt.Errorf("failed to advance page while querying Billing container for '%s': %w", resourceID, err)
+		}
+
+		for _, item := range queryResponse.Items {
```

Shouldn't a length >1 be an error here?

---

Perhaps, but then that means never recording a deletion timestamp and we would potentially continue billing a deleted cluster.

---

I'm assuming that resource ID is not the primary key since it contains the semantic name of the cluster, and someone deleting it then quickly re-creating it would lead to more than one entry in the database for this resource ID, just with different creation timestamps. Since you also `AND NOT IS_DEFINED(c.deletionTime)`, the only case where we end up with more than one item in this query is, what, a race condition where the user is deleting the previous cluster and creating the current one immediately, and we have not yet updated the previous deletion time, but have already written the new one to the DB? 

Marking the new cluster as deleted sounds like the wrong thing to do there. 

Grabbing the first (sorted how?) document and returning that billing ID also seems wrong, seems like we need to carefully reason about what happened to get us into this state. Fail loudly when N > 1 and alert something, if putting in some 404 transient is not sufficient to eventually get to the right outcome. Silently doing "something" seems like the least best option.

---

> Shouldn't a length >1 be an error here?
Addressed this much in a new commit.